File: ext/post-to-facebook.py
import json
import urllib.parse
import os
import sys
import re

# Konfigurasi Path & URL
JSON_FILE = 'artikel.json'
DATABASE_FILE = 'mini/posted-facebook.txt'
DOMAIN_URL = 'https://dalam.web.id'

def slugify(text):
    # Pastikan ada indentasi 4 spasi di bawah ini
    text = text.strip().lower()
    text = re.sub(r'\s+', '-', text)
    return text

def main():
    if not os.path.exists(JSON_FILE):
        print("Error: artikel.json tidak ditemukan")
        sys.exit(1)

    # Load data
    with open(JSON_FILE, 'r', encoding='utf-8') as f:
        data = json.load(f)

    # --- LOAD DATABASE (Cek sebagai string untuk filter Slug) ---
    posted_database = ""
    if os.path.exists(DATABASE_FILE):
        with open(DATABASE_FILE, 'r', encoding='utf-8') as f:
            posted_database = f.read()

    # Gabungkan semua kategori
    all_posts = []
    for category_name, posts in data.items():
        cat_slug = slugify(category_name)

        for post in posts:
            # Struktur: [0:judul, 1:slug, 2:image, 3:date(ISO), 4:desc]
            file_name = post[1].strip()
            file_slug = file_name.replace('.html', '').replace('/', '')

            # --- FILTER AGREGAT: Gunakan continue untuk skip ke artikel berikutnya ---
            if file_slug.startswith("agregat-20"):
                continue

            # Format URL V6.9
            full_url = f"{DOMAIN_URL}/{cat_slug}/{file_slug}"

            # CEK BERDASARKAN SLUG (Anti-Spam)
            if file_slug not in posted_database:
                all_posts.append({
                    'title': post[0],
                    'slug': file_slug,
                    'url': full_url,
                    'date': post[3],
                    'desc': post[4] or "Archive.",
                    'category': category_name
                })

    # Sorting Terbaru -> Lama
    all_posts.sort(key=lambda x: x['date'], reverse=True)

    if all_posts:
        target_post = all_posts[0]
        target_url = target_post['url']

        desc = target_post['desc']
        cat_raw = target_post['category']

        cat_hashtag = "#" + cat_raw.replace(" ", "").lower()
        hashtags = f"#fediverse #Repost #Ngopi {cat_hashtag} #Indonesia"

        full_msg = f"{desc}\n\n{target_post['title']}\n\n{hashtags}\n\n{target_url}"
        encoded_msg = urllib.parse.quote(full_msg)

        # Output untuk GitHub Actions
        if 'GITHUB_OUTPUT' in os.environ:
            with open(os.environ['GITHUB_OUTPUT'], 'a') as go:
                go.write(f"url={target_url}\n")
                go.write(f"encoded_msg={encoded_msg}\n")

        # Simpan URL sementara
        with open('/tmp/temp_new_url_facebook.txt', 'w') as f:
            f.write(target_url + '\n')

        print(f"✅ Berhasil memproses: {target_url}")
    else:
        print("✅ Tidak ada artikel baru untuk Facebook.")

if __name__ == "__main__":
    main()

File: ext/post-to-facebook.js
import fs from "node:fs";
import path from "node:path";

// Konfigurasi Path & URL
const JSON_FILE = "artikel.json";
const DATABASE_FILE = "mini/posted-facebook.txt";
const DOMAIN_URL = "https://dalam.web.id";

function slugify(text) {
  if (!text) return "";
  return String(text)
    .trim()
    .toLowerCase()
    .replace(/\s+/g, "-");
}

async function main() {
  // 1. Cek file artikel.json
  if (!fs.existsSync(JSON_FILE)) {
    console.error("❌ Error: artikel.json tidak ditemukan");
    process.exit(1);
  }

  // 2. Load data menggunakan Bun (Cepat)
  const data = await Bun.file(JSON_FILE).json();

  // 3. Load database posted
  let postedDatabase = "";
  if (fs.existsSync(DATABASE_FILE)) {
    postedDatabase = await Bun.file(DATABASE_FILE).text();
  }

  const allPosts = [];
  for (const [categoryName, posts] of Object.entries(data || {})) {
    const catSlug = slugify(categoryName);

    if (!Array.isArray(posts)) continue;
    for (const post of posts) {
      // Struktur: [0:judul, 1:slug, 2:image, 3:date(ISO), 4:desc]
      const fileName = String(post[1] || "").trim();
      const fileSlug = fileName.replace(/\.html$/i, "").replace(/\//g, "");

      if (fileSlug.startsWith("agregat-20")) continue;

      const fullUrl = `${DOMAIN_URL}/${catSlug}/${fileSlug}`;

      if (!postedDatabase.includes(fileSlug)) {
        allPosts.push({
          title: post[0] || "Untitled",
          slug: fileSlug,
          url: fullUrl,
          date: post[3] || "",
          desc: post[4] || "Archive.",
          category: categoryName,
        });
      }
    }
  }

  // Sorting Terbaru -> Lama
  allPosts.sort((a, b) => new Date(b.date) - new Date(a.date));

  if (allPosts.length > 0) {
    const targetPost = allPosts[0];

    // Persiapan Pesan
    const catHashtag = "#" + targetPost.category.replace(/\s+/g, "").toLowerCase();
    const fullMsg = `${targetPost.desc}\n\n${targetPost.title}\n\n${catHashtag}\n\n${targetPost.url}`;

    // URL Encoding untuk kebutuhan API/GitHub Actions
    const encodedMsg = encodeURIComponent(fullMsg);

    // Output untuk GitHub Actions
    if (process.env.GITHUB_OUTPUT) {
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `url=${targetPost.url}\n`);
      fs.appendFileSync(process.env.GITHUB_OUTPUT, `encoded_msg=${encodedMsg}\n`);
    }

    // Simpan log sementara (opsional)
    await Bun.write("/tmp/temp_new_url_facebook.txt", targetPost.url + "\n");

    console.log(`✅ Berhasil memproses untuk Facebook: ${targetPost.url}`);

    // Note: Jika kamu punya API Facebook Page, kamu bisa panggil di sini.
    // Jika tidak, GitHub Actions bisa menggunakan 'encoded_msg' ini
    // untuk dikirim via bot lain (seperti Messenger/Telegram bot).

  } else {
    console.log("✅ Tidak ada artikel baru untuk Facebook.");
  }
}

main();